modules/distributor/forwarder.go

```diff
@@ -0,0 +1,311 @@
+package distributor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/atomic"
+	"go.uber.org/multierr"
+)
+
+var (
+	metricForwarderPushes = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_pushes",
+		Help:      "Total number of successful requests queued up for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderDroppedPushes = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_dropped_pushes",
+		Help:      "Total number of failed pushes to the queue for a tenant to the forwarder",
+	}, []string{"tenant"})
+	metricForwarderQueueLength = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Name:      "distributor_forwarder_queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"tenant"})
+)
+
+type forwardFunc func(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) error
+
+type request struct {
+	keys   []uint32
+	traces []*rebatchedTrace
+}
+
+// forwarder queues up traces to be sent to the metrics-generators
+type forwarder struct {
+	services.Service
+
+	// per-tenant queue managers
+	queueManagers map[string]*queueManager
+	mutex         sync.Mutex
+
+	forwardFunc forwardFunc
+
+	o                 *overrides.Overrides
+	overridesInterval time.Duration
+	shutdown          chan interface{}
+}
+
+func newForwarder(fn forwardFunc, o *overrides.Overrides) *forwarder {
+	rf := &forwarder{
+		queueManagers:     make(map[string]*queueManager),
+		mutex:             sync.Mutex{},
+		forwardFunc:       fn,
+		o:                 o,
+		overridesInterval: time.Minute, // TODO: Make this configurable
+		shutdown:          make(chan interface{}),
+	}
+
+	rf.Service = services.NewIdleService(rf.start, rf.stop)
+
+	return rf
+}
+
+// SendTraces queues up traces to be sent to the metrics-generators
+func (f *forwarder) SendTraces(ctx context.Context, tenantID string, keys []uint32, traces []*rebatchedTrace) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	select {
+	case <-f.shutdown:
+		return
+	default:
+	}
+
+	if _, ok := f.queueManagers[tenantID]; !ok {
+		queueSize := f.o.MetricsGeneratorForwarderQueueSize(tenantID)
+		workerCount := f.o.MetricsGeneratorForwarderWorkers(tenantID)
+		f.queueManagers[tenantID] = newQueueManager(tenantID, queueSize, workerCount, f.forwardFunc)
+	}
+
+	err := f.queueManagers[tenantID].pushToQueue(ctx, &request{keys: keys, traces: traces})
+	metricForwarderPushes.WithLabelValues(tenantID).Inc()
+	if err != nil {
+		// TODO: we may want to log the err
```

Comment:
I wasn't sure how to handle this. The idea is that pushing request is non-blocking to the ingest path, so and error here isn't really an error (we don't want to handle it). Still, we can return to the distributor and let it handle it or keep it _hidden_ in here. I don't mind either approach.

As for logging, I don't know. It can become very spammy (?). Setting alerts for dropped pushes should be enough to trigger an operator to take a look at the queue, metrics-generators, latencies, etc. to get a feel for the problem on why they're being dropped. But, I'm open to logging failures too. I'm not very convinced with either solution.